### ISSUE TITLE
ModifyDrawData hook

### ISSUE THREAD

**Issue by stormytuna** [bluemagic123/tModLoader#3475]
### Do you intend to personally contribute/program this feature?

Yes

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

A `ModifyDrawData` hook would allow modders to make small changes to how an item/projectile/npc is drawn, instead of using the existing `PreDraw` hooks, initialising all the variables themselves and changing the small thing they want to.

The hook I'm proposing could utilise the existing `DrawData` struct Terraria provides and would exist on anything with an already existing `PreDraw` hook.

### What does this proposal attempt to solve or improve?

This will improve code quality of simpler changes to an entity's draw paramaters similar to how `ModifyShootStats` does. One example could be making an NPC shake by adding a random vector to its draw position. Currently, you would need to do something like this:
```cs
	public override bool PreDraw(SpriteBatch spriteBatch, Vector2 screenPos, Color drawColor) {
		Main.instance.LoadNPC(Type);
		Texture2D texture = TextureAssets.Npc[Type].Value;
		Vector2 drawPosition = NPC.Center - Main.screenPosition;
		drawPosition += Main.rand.NextVector2Circular(ShakeStrength, ShakeStrength);
		Rectangle sourceRect = texture.Frame();
		Vector2 origin = texture.Size() / 2f;
		SpriteEffects effects = NPC.spriteDirection == 1 ? SpriteEffects.None : SpriteEffects.FlipHorizontally;
		spriteBatch.Draw(texture, drawPosition, sourceRect, drawColor, NPC.rotation, origin, NPC.scale, effects, 0);
		
		return false;
	}
```
With a `ModifyDrawData` hook, it could look something more like this:
```cs
    public override void ModifyDrawData(ref DrawData) {
        DrawData.position += Main.rand.NextVector2Circular(ShakeStrength, ShakeStrength);
    }
```

### Which (other) solutions should be considered?

It could be called `ModifyDrawInfo` or `ModifyDrawParams`

**Comment by DavidF-Dev** [comment]
Idea:
```cs
public override void ModifyDrawData(ref EntityDrawInfo drawInfo)
{
  drawInfo.SpriteBatch = new SpriteBatch(/* ... */);
  drawInfo.Position.X += 10f;
  drawInfo.Rotation += 3.14f;
}
```

Also I think you should keep rotation and scale there! :D

**Comment by stormytuna** [comment]
Edited to clarify something and include that, thanks David :)

**Comment by direwolf420** [comment]
This is very difficult to implement due to the wide variety of ways certain things draw, i.e. in addition to patching the regular draw case, you would need to patch all the cases that check for aiStyle or similar common properties that modded entities might get drawn through (i.e. spears/drills).
Not to mention that such a hook would be nigh impossible to implement for `GlobalX` classes to affect vanilla content as plenty vanilla entities utilize manual drawing.

**Comment by Chicken-Bones** [comment]
I agree with dire, I understand how such a hook could make altering drawing easier, but I think we'd have a never-ending list of 'doesn't work for x projectile/npc/etc'
